### ISSUE TITLE
Providing using the class as token from two different module instance does not work

### ISSUE THREAD

**Issue by zenozen** [nestjs/nest#2379]
## Current behavior

I'm trying to write a shared config module, and leveraging class-validator to verify configs.
Say I have a `DatabaseSettings` class. The idea is, this module will provide the corresponding instance of this class, by reading from `process.env` and map and verify values.

When this module is imported from two different places providing two different config class, the second one will not be able to inject.

## Input Code

**some-module.settings.ts**: where the setting class is saved
```ts
export class DatabaseSettings {
  @IsString()
  @IsNotEmpty()
  @Expose({ name: 'DB_HOST' })
  readonly dbHost!: string;
}

export class OtherSettings {
  @IsString()
  @Expose({ name: 'OTHER_HOST' })
  readonly otherHost!: string;
}
```

**settings.module.ts**
```ts

dotenv.config();

@Module({})
export class SettingsModule {
  static forSettings(...classes: Array<Type<any>>): DynamicModule {
    const settingProviders: Provider[] = classes.map(clazz => ({
      provide: clazz,
      useFactory: async () => {
        const value = plainToClass(clazz, process.env, {
           enableImplicitConversion: true,
           excludeExtraneousValues: true,
        });
        const validationErrors = await validate(value);
        if (validationErrors.length > 0) {
          throw new Error('Oops');
        }
        return value;
      },
    }));
    return {
      module: SettingsModule,
      providers: settingProviders,
      exports: settingProviders,
    };
  }
}
```

**app.module.ts**
```ts
@Module({
  imports: [
    TypeOrmModule.forRootAsync({
      imports: [SettingsModule.forSettings(DatabaseSettings)],
      useFactory: dbSettings => ...,
      inject: [DatabaseSettings],
    }),
    AuthModule,
    ...
  ]
})
export class AppModule {}
```

**auth.module.ts**
```ts
@Module({
  imports: [
    SettingsModule.forSettings(OtherSettings),
    ...
  ],
  providers: [ XService ], // Use OtherSettings. This will fail saying XService's dependency (OtherSettings) cannot be found
  ...
})
export class AuthModule {}
```

## Expected behavior
It should work fine.

## Possible Solution
I found two possible solutions.

1. We can work around this by providing a string token rather than the class itself, for example,
`__config:DatabaseSettings` instead of `DatabaseSettings` the class, and have a separate inject decorators like `@InjectConnection` in typeorm or `@InjectConfig` in nest-config, but it's very verbose, like
```ts
constructor(
  // Oops, repeated myself
  @InjectSettings(DatabaseSettings) databaseSettings: DatabaseSettings
)

// And it doesn't seem to work properly without calling the token creation function in this module imports
TypeOrmModule.forRootAsync({
  imports: [SettingsModule.forSettings(DatabaseSettings)],
  useFactory: dbSettings => ...,
  inject: [createSettingsToken(DatabaseSettings)], // createSettingsToken has to be exposed?
})
```

FWIU, typeorm expose those decorators since custom arguments can be used, but in our case, there's no customization going on.

2. I tracked down to where module token is created. 

While creating this token by serialize `dynamicModuleMetadata` (which has all the imports, exports, providers in it), `fast-safe-stringify` ignores all the class/functions and replace them with `null`. Hence the two module created above by two separate `forSettings` are actually the same hash, because their dynamic metadata are both serialized to`'{"providers":[null,{"inject":[null]}],"exports":[{"inject":[null]}]}'`. The second injection would fail because it found the cached first module using this hash.

https://github.com/nestjs/nest/blob/master/packages/core/injector/module-token-factory.ts#L28

I don't know if this is intentional (it certainly looks not, having all these nulls in there is weird), but adding a replacing function seems to fix it.

```ts
  public getDynamicMetadataToken(
    dynamicModuleMetadata: Partial<DynamicModule> | undefined,
  ): string {
    function replacer(key: string, value: any) {
      if (typeof value === 'function') {
        return value.name;
      } else {
        return value;
      }
    }
    // Uses safeStringify instead of JSON.stringify
    // to support circular dynamic modules
    return dynamicModuleMetadata ? stringify(dynamicModuleMetadata, replacer) : '';
  }
```

Here's an example of serialized `dynamicModuleMetadata` before and after.
```
Before: 

{"imports":[{"providers":[null,{"inject":[null]}],"exports":[{"inject":[null]}]}],"providers":[{"provide":"TypeOrmModuleOptions","inject":["__config:DatabaseSettings"]},{"inject":[null]},{"inject":["TypeOrmModuleOptions"]},{"provide":"TypeOrmModuleId","useValue":"62a69bc5-fa76-46ac-b981-e7c5122a9339"}],"exports":[{"inject":[null]},{"inject":["TypeOrmModuleOptions"]}]}

After (looks much saner):

{"imports":[{"module":"SettingsModule","providers":["SettingsLoader",{"provide":"DatabaseSettings","useFactory":"useFactory","inject":["SettingsLoader"]}],"exports":[{"provide":"DatabaseSettings","useFactory":"useFactory","inject":["SettingsLoader"]}]}],"providers":[{"provide":"TypeOrmModuleOptions","useFactory":"useFactory","inject":["__config:DatabaseSettings"]},{"provide":"EntityManager","useFactory":"useFactory","inject":["Connection"]},{"provide":"Connection","useFactory":"useFactory","inject":["TypeOrmModuleOptions"]},{"provide":"TypeOrmModuleId","useValue":"f8f827e7-177b-45f8-9ce0-accba08297d5"}],"exports":[{"provide":"EntityManager","useFactory":"useFactory","inject":["Connection"]},{"provide":"Connection","useFactory":"useFactory","inject":["TypeOrmModuleOptions"]}]}
```

## Environment

<pre><code>
Nest version: 6.2.4
 
For Tooling issues:
- Node version: 12.4.0
- Platform:  Mac

Others:
</code></pre>

**Comment by maksimkurb** [comment]
I think that I faced the same issue with module injection by token.
I made a [sample repro](https://github.com/maksimkurb/nest-httpmodule-bug/tree/another-bug)

**Comment by zenozen** [comment]
@maksimkurb Thanks. I tried your repro, the app seems to start fine?

**Comment by maksimkurb** [comment]
@zenozen, please switch to `another-bug` branch 😃

**Comment by zenozen** [comment]
Ah, right. 

Did some debugging, it is the same issue in some sense (or just intentional...), because when importing the CommonModule, the metadata for those two CommonModule are both `'{"providers":[{"provide":"CommonModule/CommonService"}],"exports":[{"provide":"CommonModule/CommonService"}]}'`. Info about what's actually provided is missing, hence resulting in the same hash.

**Comment by maksimkurb** [comment]
@kamilmysliwiec, can you look into it, please?
zenozen's possible solution resolves issue in my case.

Metadata without fix:
```js
// the same, as zenozen stated
{"providers":[{"provide":"CommonModule/CommonService"}],"exports":[{"provide":"CommonModule/CommonService"}]}
{"providers":[{"provide":"CommonModule/CommonService"}],"exports":[{"provide":"CommonModule/CommonService"}]}
```

and with this fix:
```js
// has useClass field
{"providers":[{"provide":"CommonModule/CommonService","useClass":"RealCommonService"}],"exports":[{"provide":"CommonModule/CommonService","useClass":"RealCommonService"}]}
{"providers":[{"provide":"CommonModule/CommonService","useClass":"DemoCommonService"}],"exports":[{"provide":"CommonModule/CommonService","useClass":"DemoCommonService"}]}
```

**Comment by kamilmysliwiec** [comment]
@zenozen would you like to create a PR for this issue? :)

**Comment by zenozen** [comment]
@kamilmysliwiec Here it is :)

I can't seem to get the integration tests running, so not sure if those are broken...
Let me know if anything needs changes, it looks like a trivial change

**Comment by kamilmysliwiec** [comment]
Thanks again! Published as 6.4.0

**Comment by zenozen** [comment]
Upgraded to 6.4.0 and it works! Thanks for merging and publishing, my code looks pretty again! :)
